dataRetriever/resolvers/topicResolverSender/topicResolverSender.go

```diff
@@ -193,9 +232,56 @@ func (trs *topicResolverSender) sendOnTopic(peerList []core.PeerID, topicToSendR
 	}
 	log.Trace("requests are sent to", logData...)
 
+	log.Debug("request peers histogram", "max peers to send", maxToSend, "topic", topicToSendRequest, "histogram", histogramMap)
+
 	return msgSentCounter
 }
 
+func (trs *topicResolverSender) getPreferredPeer(pType peerType) core.PeerID {
+	peersInShard, found := trs.getPeersBasedOnPeerType(pType)
+	if !found {
+		return ""
+	}
+
+	indexes := make([]int, 0, len(peersInShard))
+	for idx := range peersInShard {
+		indexes = append(indexes, idx)
+	}
+
+	shuffledIndexes := random.FisherYatesShuffle(indexes, trs.randomizer)
+
+	return peersInShard[shuffledIndexes[0]]
+}
+
+func (trs *topicResolverSender) getPeersBasedOnPeerType(pType peerType) ([]core.PeerID, bool) {
```

Comment:
you can directly request preferred peers for the targetShardID and then directly get from preferredPeers[targetShardID]

Reply:
refactored this way